containers/podman-desktop
# feat: collect application logs and diagnostic data

### What does this PR do?

The following changes proposal adds an ability to collect application logs and various diagnostic data, e.g. Docker & Podman versions, os info, application preferences, display configuration, etc. Also it provides logging mechanism for the backend side.

By default all application logs now stored in:

- macOS:
  - `$HOME/Library/Logs/Podman Desktop`
- Linux:
  - `~/.config/Podman Desktop/logs`
- Windows:
  - `%APPDATA%\Podman Desktop\logs`

---

It provides two API interfaces:

- `DiagnosticInfoProvider`
```typescript
  export interface DiagnosticInfoProvider {
     title: string;
     collectInfo(): Promise<string>;
   }
```

`DiagnosticInfoProvider` provides any additional information like version, paths, snapshots, etc. which will be displayed in `diagnostic-information.txt` file. The default implementation for [Podman](https://github.com/containers/podman-desktop/blob/eec6e2192cd91b34d4ed1cca2eb3444a3d182097/extensions/podman/src/diagnostic/podman-diagnostic-info-provider.ts) and [Docker](https://github.com/containers/podman-desktop/blob/eec6e2192cd91b34d4ed1cca2eb3444a3d182097/extensions/docker/src/diagnostic/docker-diagnostic-info-provider.ts) retrieves the information from `/version` api call.

- `DiagnosticLogsProvider`
```typescript
  export interface DiagnosticLogsProvider {
     title: string;
     collectLogs(): Promise<string[]>;
   }
```

`DiagnosticLogsProvider` provides paths to the logs file that serves particular extension. For example it can be a path to crc log file, which will be included into general archive. `title` field is responsible for how the folder with provided logs will be named.

In Extension API there is a new namespace provided with name `diagnostic`, which contains all necessary methods to register and unregister above providers:

```typescript
  export namespace diagnostic {
     export function registerDiagnosticInfoProvider(diagnosticInfoProvider: DiagnosticInfoProvider): Disposable;
     export function unregisterDiagnosticInfoProvider(diagnosticInfoProvider: DiagnosticInfoProvider): void;
     export function registerDiagnosticLogsProvider(diagnosticLogsProvider: DiagnosticLogsProvider): Disposable;
     export function unregisterDiagnosticLogsProvider(diagnosticLogsProvider: DiagnosticLogsProvider): void;
   }
```

calling Disposable is the same as calling unregisterDiagnosticInfoProvider for the particular provider.

It also allow to process a command line arguments:

- `--logLevel=[error|warn|info|verbose|debug|silly]`
  - to put into logs messages not higher than given log level

- `--logPath=[.../path/to/log/folder]`
  - to override the default log folder location

### Screenshot/screencast of this PR

macOS:
![macOS](https://user-images.githubusercontent.com/1968177/236249856-334305cc-0be8-4459-aff3-65daa7a5cc85.jpg)

Windows:
![windows](https://user-images.githubusercontent.com/1968177/236249906-6c0982f9-52ce-45f9-952d-71f5d1a49967.jpg)

Fedora:
![Снимок экрана от 2023-05-04 18-05-22](https://user-images.githubusercontent.com/1968177/236249934-9e4ba8d7-4b6e-421a-b518-46447a0f8949.png)

### What issues does this PR fix or reference?

https://github.com/containers/podman-desktop/issues/1423

### How to test this PR?

Scenario 1:

- Download asset for particular OS
- Open menu item `Help` -> `Collect Logs and Diagnostic Data`
- Agree with the dialog and press button `Show in Finder`/`Show in File Explorer`
- Navigator will open the folder with selected zip archive
- Extract the newly created zip archive and check if it contents sub folder `Podman Desktop` with application logs and file `diagnostic-information.txt` with some system diagnostic information

Scenario 2:

- Download asset for particular OS
- Enable in Preferences option `Hide The Dialog` in `Diagnostic` section
- Open menu item `Help` -> `Collect Logs and Diagnostic Data`
- Navigator will open the folder with selected zip archive without any dialogs
- Extract the newly created zip archive and check if it contents sub folder `Podman Desktop` with application logs and file `diagnostic-information.txt` with some system diagnostic information
